api/src/main/java/com/datatorrent/api/StorageAgent.java

```diff
@@ -94,4 +96,31 @@
     public void setApplicationAttributes(AttributeMap map);
   }
 
+  /**
+   * Storage agent which can take checkpoints asynchronously.
```

Comment:
It is not clear what someone who wants to implement a new async storage agent implementation needs to do. I think to start with, it is confusing that they would need to implement both the StorageAgent and the AsyncStorageAgent interface. Then it needs to be documented that they need to the operation in two steps. First, a store method where they need to make a temporary copy of the given object and then a finalize method where they can store it.

The same store method seems to have two different meanings when it comes to sync case vs async case. In sync case it is supposed to completely save the checkpoint. In async case it is supposed to only make a temporary copy. This can be confusing to the user.

I suggest for these reasons keep AsyncStorageAgent as an internal interface.